### PR TITLE
KAFKA-13504: Retry connect internal topics' creation in case of InvalidReplicationFactorException

### PR DESCRIPTION
**Pull request by akatona84**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by mimaison** [comment]
Thanks @akatona84 for the PR.
In order to add new configurations, we need to have a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals)

**Comment by akatona84** [comment]
> Thanks @akatona84 for the PR. In order to add new configurations, we need to have a [KIP](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals)

Hey, thx for checking. Any idea/suggestions avoiding (too much) new configs because of the retry?

**Comment by mimaison** [comment]
Sorry for the delay, but this finally made it back to the top of my review queue.

I'm not sure we necessarily need to do something here. Starting Connect from scratch when not enough brokers are running must be a pretty rare scenario (maybe more common when developing). Connect has no idea when, or even if, enough brokers will join the cluster so it's hard to decide whether to keep retrying to just stop and let the environment (for example k8s) handle the restarts.

If you definitively want Connect to do some retries, maybe we could retry for `default.api.timeout.ms`. I typically find setting a bound as a duration more intuitive than setting a number of retries. Also having separate retries settings for each topic may be a bit overkill :)

**Comment by akatona84** [comment]
I totally agree that I have too many configs :D

Yeah, it's happening quite often when testing, just starting the brokers and connect cluster and the health check of connect alerts due to its unexpected death.
I know it should be the orchestrator's duty to start connect after brokers are fine, yet thought would be nice to make it more robust.
Now I tried to rebase and see KIP-618 efforts on trunk, will try again to rebase and rewrite it to time bound retries. It would still need two configs: the timeout and the backoff as I see default.api.timeout.ms is not defined for the worker/distributed.

tbh I'm not 100% sure if this robustness worth the effort, I'll think about it.

**Comment by akatona84** [comment]
@urbandan , @lhunyady I've reworked the change, now it has the timeout exception and timeout based retries (instead of retry counts).
Could you guys take a look please?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 1]
```diff
@@ -329,6 +334,48 @@ public Set<String> createTopics(NewTopic... topics) {
...
 
+    /**
+     * Implements a retry logic around creating topic(s) in case it'd fail due to InvalidReplicationFactorException
+     *
+     * @param topicDescription
+     * @param config
+     * @param time
+     * @return the same as {@link TopicAdmin#createTopics(NewTopic...)}
+     */
+    public Set<String> createTopicsWithRetry(NewTopic topicDescription, WorkerConfig config, Time time) {
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: I would avoid passing a full WorkerConfig here, and try to minimalize the amount of data passed here by only propagating the timeout and the backoff. That simplifies testing, too.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 2]
```diff
@@ -329,6 +334,48 @@ public Set<String> createTopics(NewTopic... topics) {
...
+                    if (retryableTopicCreationException(e)) {
+                        log.info("'{}' topic creation failed due to '{}', retrying, {}ms remaining",
+                                topicDescription.name(), e.getMessage(), timer.remainingMs());
+                        timer.sleep(backOffMs);
+                    } else {
+                        throw e;
+                    }
+                }
+            } while (timer.notExpired());
+            throw new IllegalStateException("This should not happen");
```
Comment (lhunyady) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: This should happen if the time expires right?

Comment (akatona84) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: yeah, thx!
forgot to transform it to timeout exception after the retry based logic rewirte.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java (hunk 1 above)
Comment (akatona84) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: This was exactly what I don't like in my change :D 
I was trying to avoid duplicating the code for this instanceof check. I'll find another place.

**Comment by mimaison** [comment]
@akatona84 Thanks for taking another look and simplifying the configuration, I think it looks much better. This will still need a KIP.

**Comment by viktorsomogyi** [comment]
@akatona84 would you consider using `default.api.timeout.ms` and backoff based on `retry.backoff.ms`? I can't really justify having these new configs introduced specificly for this until we realize we really need them. At that point someone can create a KIP.

**Comment by akatona84** [comment]
@mimaison , @viktorsomogyi I modified the change as you suggested, I'm curious how you like it this way.
sorry for the loong delay

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBackingStore.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.kafka.connect.util.TopicAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public abstract class KafkaTopicBackingStore {
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBackingStore.java]: nit: KafkaTopicStore or KafkaTopicBackedStore would be more descriptive

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBackingStore.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    private static final Logger log = LoggerFactory.getLogger(KafkaTopicBackingStore.class);
+
+    Consumer<TopicAdmin> topicInitializer(String topic, NewTopic topicDescription, WorkerConfig config, Time time) {
+        return admin -> {
+            log.debug("Creating admin client to manage Connect internal {} topic", getTopicPurpose());
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBackingStore.java]: nit: I know this was copied over from the old method, but this doesn't seem accurate, can you please fix it?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 5]
```diff
@@ -328,6 +331,41 @@ public Set<String> createTopics(NewTopic... topics) {
...
+            timer.sleep(backOffMs);
+        } while (timer.notExpired());
+        throw new TimeoutException("Timeout expired while trying to create topic(s)");
+    }
+
+    private boolean retryableTopicCreationException(ConnectException e) {
+        // createTopics wraps the exception into ConnectException
+        // which can be an ExecutionException from future get which was caused by InvalidReplicationFactorException
+        // or can be a TimeoutException
+        return e.getCause() instanceof ExecutionException && e.getCause().getCause() instanceof InvalidReplicationFactorException
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: wondering if this should be more generic, e.g. InvalidReplicationFactorException can be at any place in the exception chain
createTopics only documents ConnectException, and nothing else, so a future commit can freely change the chain of exceptions as long as the topmost is ConnectException

**Review by urbandan — APPROVED** [review]
LGTM, thanks for the change!

**Review by urbandan — CHANGES_REQUESTED** [review]
sorry, realized I have 1 pending comment

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 6]
```diff
@@ -360,10 +361,20 @@ public Set<String> createTopicsWithRetry(NewTopic topicDescription, long timeout
...
     private boolean retryableTopicCreationException(ConnectException e) {
         // createTopics wraps the exception into ConnectException
-        // which can be an ExecutionException from future get which was caused by InvalidReplicationFactorException
+        // to retry the creation, it should be an ExecutionException from future get which was caused by InvalidReplicationFactorException
         // or can be a TimeoutException
-        return e.getCause() instanceof ExecutionException && e.getCause().getCause() instanceof InvalidReplicationFactorException
-                || e.getCause() instanceof TimeoutException;
+        List<Class<? extends Exception>> causesToRetry = Arrays.asList(
+                InvalidReplicationFactorException.class,
+                TimeoutException.class);
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: wondering if it makes sense to check for TimeoutException with the latest config changes - if the admin client itself encounters a timeout, and we use the same timeout config, this is probably not necessary anymore

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 7]
```diff
@@ -360,10 +361,20 @@ public Set<String> createTopicsWithRetry(NewTopic topicDescription, long timeout
 
     private boolean retryableTopicCreationException(ConnectException e) {
         // createTopics wraps the exception into ConnectException
-        // which can be an ExecutionException from future get which was caused by InvalidReplicationFactorException
+        // to retry the creation, it should be an ExecutionException from future get which was caused by InvalidReplicationFactorException
         // or can be a TimeoutException
-        return e.getCause() instanceof ExecutionException && e.getCause().getCause() instanceof InvalidReplicationFactorException
-                || e.getCause() instanceof TimeoutException;
+        List<Class<? extends Exception>> causesToRetry = Arrays.asList(
```
Comment (urbandan) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: nit: should be a constant

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 8]
```diff
@@ -328,6 +335,48 @@ public Set<String> createTopics(NewTopic... topics) {
         return createOrFindTopics(topics).createdTopics();
     }
 
+    /**
+     * Implements a retry logic around creating topic(s) in case it'd fail due to InvalidReplicationFactorException
```
Comment (viktorsomogyi) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: nit: add "or TimeoutException"

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 9]
```diff
@@ -328,6 +335,48 @@ public Set<String> createTopics(NewTopic... topics) {
...
+     *
+     * @param topicDescription
+     * @param timeoutMs
+     * @param backOffMs
+     * @param time
+     * @return the same as {@link TopicAdmin#createTopics(NewTopic...)}
```
Comment (viktorsomogyi) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java]: nit: please fill out parameters and return value if you add javadocs.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 10]
```diff
@@ -70,6 +74,9 @@
 public class TopicAdmin implements AutoCloseable {
 
     public static final TopicCreationResponse EMPTY_CREATION = new TopicCreationResponse(Collections.emptySet(), Collections.emptySet());
+    private static final List<Class<? extends Exception>> CAUSES_TO_RETRY_TOPIC_CREATION = Arrays.asList(
+            InvalidReplicationFactorException.class,
+            TimeoutException.class);
```
Comment (viktorsomogyi) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:79]: Since TimeoutException is a RetriableException I was wondering whether we could refactor InvalidReplicationFactorException because in some circumstances it can be retriable and I think even if there won't be more brokers, it doesn't hurt too much to retry a few times with backoff. What do you think @mimaison, would this be considered an API change?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java [hunk 11]
```diff
@@ -328,6 +335,48 @@ public Set<String> createTopics(NewTopic... topics) {
...
+    /**
+     * Implements a retry logic around creating topic(s) in case it'd fail due to InvalidReplicationFactorException
+     *
+     * @param topicDescription
+     * @param timeoutMs
+     * @param backOffMs
+     * @param time
+     * @return the same as {@link TopicAdmin#createTopics(NewTopic...)}
+     */
+    public Set<String> createTopicsWithRetry(NewTopic topicDescription, long timeoutMs, long backOffMs, Time time) {
```
Comment (viktorsomogyi) [inline connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:349]: I think it would be a bit more robust to add an overload to {{org.apache.kafka.connect.util.RetryUtil#retryUntilTimeout}} that specifies a set of exceptions or a condition to retry on and use that here.

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java (hunk 10 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:79]: IMO this would be a bit too drastic; the fallout of a change like that scales linearly with the retry duration that the user has configured and may delay surfacing valid issues for a bit too long. It could be especially frustrating in quickstart scenarios where someone defines a replication factor that exceeds the number of local brokers and has to wait an entire minute to find out what the problem is.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+import org.apache.kafka.connect.util.TopicAdmin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+import java.util.function.Supplier;
+
+public abstract class KafkaTopicBasedBackingStore {
```
Comment (C0urante) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java:36]: Is inheritance necessary here? It seems like we might make this a standalone class that can be composed into the various `Kafka*BackingStore` classes (and then possibly others) instead of only making it available through subclassing.

(Not a blocker)

Comment (akatona84) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java:36]: I tried it.
in one hand:
It'd require the composed field to be static, because of static methods would use the topicInitializer method:
https://github.com/apache/kafka/blob/15dfe065fb0ef34145ec204123c592e468842de8/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java#L129

in the other hand (which relates to the first one a bit): I'm not sure if I could inject this composition nicely, so that I can test it (or mock part of it) easily.
Currently the tests are prepared in a way that the `createKafkaBasedLog` is the method of the actual backingstore class, it would need a quite bit of a refactor on the tests too if I want to avoid this inheritance :)

wdyt?

File: connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java (hunk 11 above)
Comment (akatona84) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:349]: That retrier is constructed for actually retriable exception and in our case the exception is wrapped in another exception and don't really want to touch/refactor/hack this utility method to handle these kind of exceptions too and adding more not-necessarily retryable exceptions to check in the condition where it retries.
This refactor would change the semantics of the retries where it's being used currently.

long story short, I feel this topic creation retry is a bit special and wouldn't been used this way in many places later on.

Comment (viktorsomogyi) [reply connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:349]: Ok, that's fair.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java (hunk 12 above)
Comment (C0urante) [reply connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaTopicBasedBackingStore.java:36]: Ah, I forgot about the static constructors for the offset backing store. Yeah, probably not worth it to add another constructor variant for everything.

I was thinking about isolating this logic into a `Supplier<KafkaBasedLog>`, and we could then have something like:

```java
public class LogBuilder {
  public static Supplier<KafkaBasedLog<K, V>> createLog(String topic, String topicConfig, String topicPurpose...) {
    return () -> {
      KafkaBasedLog<K, V> result = ...
      // create and set up KafkaBasedLog, including performing topic initialization here
      return result;
    }
  }
}

public class KafkaConfigBackingStore implements ConfigBackingStore {
  public KafkaConfigBackingStore(Supplier<KafkaBasedLog> setupAndCreateKafkaBasedLog, ...)
}
```

Which I think would still technically work here (and honestly be a bit cleaner than how we're hacking the `setupAndCreateKafkaBasedLog` method in tests right now), but given the number of new constructors it would require (which introduce their own amount of ugliness), we can stick with what's in this PR right now.

**Comment by akatona84** [comment]
test this please

**Comment by akatona84** [comment]
Does that kip label still apply?

**Review by urbandan — APPROVED** [review]
LGTM, thanks for the change!

**Comment by akatona84** [comment]
rebased

**Comment by viktorsomogyi** [comment]
@akatona84 seems like there is a compile issue. Please fix it.

**Comment by akatona84** [comment]
thx. rebased again and did some adjustments. (this time I built it and ran my modified unit tests as well :) )

**Review by viktorsomogyi — APPROVED** [review]
Seems like https://issues.apache.org/jira/browse/KAFKA-12384 fails here too based on that the assertion error is failing with the same parameters. Other flakies were successful on a rerun locally.
Will merge it tomorrow (just didn't want to commit and leave).

**Comment by viktorsomogyi** [comment]
Thank you @akatona84 for the contribution!
